LdapForNet/LdapEntry.cs

```diff
@@ -1,287 +1,352 @@
+        }
+
+        public DateTime? GetWhenChanged()
+        {
+            var whenChanged = GetString(LdapAttributes.WhenChanged);
+            if (whenChanged != null)
+            {
+                var date = DateTime.ParseExact(whenChanged, "yyyyMMddHHmmss.f'Z'", CultureInfo.InvariantCulture,
+                    DateTimeStyles.AssumeLocal);
+                return DateTime.SpecifyKind(date, DateTimeKind.Utc);
+            }
+
+            return null;
+        }
+
+        public DateTime? GetModifyTimestamp()
+        {
+            var modifyTimestamp = GetString(LdapAttributes.ModifyTimestamp);
+            if (modifyTimestamp != null)
+            {
+                var date = DateTime.ParseExact(modifyTimestamp, "yyyyMMddHHmmss'Z'", CultureInfo.InvariantCulture,
+                    DateTimeStyles.AssumeLocal);
+                return DateTime.SpecifyKind(date, DateTimeKind.Utc);
+            }
+
+            return null;
+        }
+
+        public IEnumerable<string> GetMemberOf() => GetStrings(LdapAttributes.MemberOf);
+
+        public UserAccountControl GetUserAccountControl()
+        {
+            var attribute = GetString(LdapAttributes.UserAccountControl);
+            return attribute == null ? UserAccountControl.NONE : (UserAccountControl)int.Parse(attribute);
+        }
+
+        public int GetPrimaryGroupID()
+        {
+            var attribute = GetString(LdapAttributes.PrimaryGroupID);
+            return attribute == null ? 0 : int.Parse(attribute);
+        }
+
+        public int GetUserPrimaryID()
+        {
+            var objectSid = GetAttribute(LdapAttributes.ObjectSid)?.GetValue<byte[]>();
+            if (objectSid != null)
+                return BitConverter.ToInt32(objectSid, objectSid.Length - 4); //last 4 bytes are primary group id
+
+            return -1;
+        }
+
+        public string GetString(string attributeName) => GetAttribute(attributeName)?.GetValue<string>();
+
+        public byte[] GetBytes(string attributeName) => GetAttribute(attributeName)?.GetValue<byte[]>();
+
+        public IEnumerable<string> GetStrings(string attributeName) => GetAttribute(attributeName)?.GetValues<string>() ?? Enumerable.Empty<string>();
+
+        public IEnumerable<byte[]> GetByteArrays(string attributeName) => GetAttribute(attributeName)?.GetValues<byte[]>() ?? Enumerable.Empty<byte[]>();
+    }
+
+    public class LdapModifyEntry
+    {
+        public string Dn { get; set; }
+        public List<LdapModifyAttribute> Attributes { get; set; }
+    }
+
+    public class LdapModifyAttribute
+    {
+        public string Type { get; set; }
+        public List<string> Values { get; set; }
+
+        public Native.Native.LdapModOperation LdapModOperation { get; set; } =
+            Native.Native.LdapModOperation.LDAP_MOD_REPLACE;
+    }
+
+    public class DirectoryAttribute
+    {
+        private List<byte[]> _byteValues;
```

Comment:
you could initiailize _byteValues and _stringValues with empty list

Reply:
It is better to no initialize them to save memory. If we retrieve a lot of entries from an LDAP, we will instanciate a lot of DirectoryAttribute. So it is better to only initialize our lists when we need it